### ISSUE TITLE
Simplify the use of `base` in the lobster configuration

### ISSUE THREAD

**Issue by annawoodard** [matz-e/lobster#143]
We should simplify the use of the `base` path in storage configurations. Consider this configuration:
```
storage:
    base: /store/user/einstein/foo/bar
    local: /hadoop/store/user/einstein/foo/bar
    output: "chirp://earth.crc.nd.edu:9099/foo/bar"
```

This does not work because of the call to StorageElement.test() [here](https://github.com/matz-e/lobster/blob/master/lobster/se.py#L327). One way to fix this is to change the path passed to StorageElement.test() for Chirp() objects to remove the server part, is that the route we want to go? Or should the user use, e.g: `base: "chirp://earth.crc.nd.edu:9099//"`?

In principle, if only the `local` and `output` paths are given, they should be sufficient to completely specify the configuration. If possible we should eliminate the need for the user to specify `base` at all in such cases, unless it is providing unique information (for example, the user provides only `base` and `site`.) If we can't do that, we should clarify the documentation regarding the use of `base` (currently we only describe it in the context of configurations using the `site` identifier.)

## EDIT
We are almost finished with necessary modifications on the `se-tweaks` branch. As I see it (please add anything missing) there are two remaining items to finish this up:
- [x] finish testing 
    - [x] MC [Anna + Matthias]
    - [x] slimming, input from DBS [Matthias]
    - [x] slimming, local input [Anna]
- [x] update documentation
    - [x] README [Matthias]
    - [x] example configurations (ND examples) [Anna]
    - [x] example configurations (CERN examples) [Matthias]

**Comment by klannon** [comment]
I feel like I'm not really understanding this issue.

Is it true that there are really only two types of storage: input and 
output?  Is there any type of operation that can't be classified as one 
of those two?

Is the role of "base" and "local" to try to deal with the idea that for 
those storage locations (I'm guessing mainly "output") there might be 
more than one route to reach the storage?  For example, the workers 
would have to use chirp or xrootd, but the master would presumably 
prefer to use direct access when available?

I guess I'm trying to understand what base, local, output, and input are 
really used for.  For example, if there are really only input and output 
storage locations, and the rest is just machinery to help specify the 
different ways different pieces are supposed to access the storage, why 
not reorganize so that you simply have two storage lists (input and 
output) and for each, you give a list of possible "routes" to the 
storage that any part of Lobster (master, worker, etc.) will try one at 
a time, falling back as each method fails? Something like:

storage:
    output: 
["/hadoop/store/user/einstein/foo/bar","chirp://earth.crc.nd.edu:9099/foo/bar","srm: 
//ndcms.crc.nd.edu:8443/srm/v2/server?SFN=/hadoop/store/user/einstein/foo/bar"]

Or whatever the proper YAML syntax is.  Probably something with "-", right?

**Comment by annawoodard** [comment]
I really like this idea, I think it is a little more user-friendly than what we have. It has the added benefit of allowing the user to specify the order of precedence (including having both Chirp and XrootD options for input, or output, or both!), rather than have us specify it behind-the-scenes in a way that is not transparent to the user. There would be one complication when dealing with SRM stageout for which the user can currently simply specify, for example `site: T3_US_NotreDame` and `base: /store/user/foo/bar`. Kevin had the idea that we could invent a syntax for this, maybe something like `site:T3_US_NotreDame:/foo/bar`, and add it to the list with the other options. We could do the same for native HDFS access: `hadoop: /store/user/foo/bar` which could be added to the list for input or output or both. This would be different than how we are treating native HDFS access now because we try to determine a single mapping with the `hadoop` path and apply it for both input and output.

**Comment by klannon** [comment]
First +1 on the "hadoop:" option!  (Or would it be "hdfs:/store/user/foo/bar"?)

I wouldn't treat "site" as a specific protocol.  Instead, I would use site names as a shorthand part of the url (uri?  I'm never sure...) in whatever protocol you're trying to use.  So, the standard usage would be something like "srm:T3_US_NotreDame:/store/user/einstein/foo/bar" instead of "srm://ndcms.crc.nd.edu:8443/srm/v2/server?SFN=/hadoop/store/user/" and Lobster would know to expand "T3_US_NotreDame" to be "ndcms.crc.nd.edu:8443/srm/v2/server?SFN=/hadoop".  In another case however, (presumably for input) you might want to have "root:T3_US_NotreDame:/store/user/einstein/foo/bar," etc.  I would also make it easy on yourself to pick out the T3 shorthand for expansion, so if it helps to have some kind of delimiter in the URL (e.g. [T3_US_NotreDame] or $T3_US_NotreDame$), I'd say go for it.

**Comment by matz-e** [comment]
I like this!  That would be a lot more clear than what I came up with!  I think as I had it originally implemented, there was an ordering to the transfer/access methods, which probably would still be the case (we only pass URLs to the worker now.)

I would use `srm://T3_US_NotreDame/store/user/einstein/foo/bar` as a short-hand: it's easy to match and pretty much in line with the standard URL notation.  I think it still makes sense to also keep a `site://T3_US_NotreDame/…` around that sets both SRM and XrootD access correctly.

**Comment by annawoodard** [comment]
> I like this! That would be a lot more clear than what I came up with! I think as I had it originally implemented, there was an ordering to the transfer/access methods, which probably would still be the case (we only pass URLs to the worker now.)

Yes-- I think we want to pass the worker exactly what the master has (if local is preferred on the master, it can be preferred on the worker, etc.)

> I would use srm://T3_US_NotreDame/store/user/einstein/foo/bar as a short-hand: it's easy to match and pretty much in line with the standard URL notation. 

Sounds good!

> I think it still makes sense to also keep a site://T3_US_NotreDame/… around that sets both SRM and XrootD access correctly.

My vote would be to avoid one -> many mappings if possible. Just writing out two paths isn't so much work that it's worth the possible confusion; besides, it's preferable to have the ordering be explicit.

**Comment by klannon** [comment]
I agree with @annawoodard.  I feel like one of the challenges to the previous approach was we tried to put too much logic into the code regarding how to resolve possible access methods.  If site://T3_US_... can be resolved more than one way, Lobster needs to guess what should be done.  I'd rather force the user to tell us what they want and keep the internal logic simple.

Nonetheless, I can see attraction of having shorthands.  How about this for a compromise?  Begin the implementation without the site: shorthand (or any others).  Once we have the basic approach implemented, and get some operational experience, if we see an emerging pattern (e.g. storage elements that always use xrootd for input and srm for output) we can implement a shorthand to make that pattern easy to use.

**Comment by matz-e** [comment]
I still prefer the shorthand:  if you make the copy-and-paste argument for every possible configuration option, it becomes quite unwieldy.  The site option also assumes that XrootD and SRM are present, so maybe being more explicit there helps point out limitations/invalid configurations?

I think I coded it the way it is because we're mixing 3 access methods here: the master, worker input and output.  So breaking it into input/output selects which method for the master to use? I think I've defaulted to using output settings, but it seems not logical to have a `file://` or `hdfs://` in there?  OTOH `file://` would be used to specify WQ transfers, so that's good!

Just to stress it again, the preference chain is only for the master, the workers always use the "latest" setting (SRM/xrootd trumps all.)

TL;DR: I'm OK with using the dev path you describe @klannon.  I could hack this up fairly quickly, too.

**Comment by klannon** [comment]
TL;DR only works if you put it first.  :)

I think we should talk about this "in person" (perhaps after the Thursday meeting if you'll be there.  I'm not sure we're fully on the same page yet.

For example, I was thinking that the input/output lists we're talking about would apply to both master and worker.  Talking with @annawoodard, we couldn't think of a scenario in which you'd want to specify a different preference list for the workers than the master.

**Comment by annawoodard** [comment]
> I still prefer the shorthand: if you make the copy-and-paste argument for every possible configuration option, it becomes quite unwieldy.

I still think copy-and-paste effort is lower in the difficulty hierarchy, but let's go with your preference here and if it's a big deal maybe we can tweak it later.

> I think I coded it the way it is because we're mixing 3 access methods here: the master, worker input and output. So breaking it into input/output selects which method for the master to use?

No-- why not have the master and the worker treat access the same? I think that's much clearer from the user perspective. 

>  I think I've defaulted to using output settings,

Defaulting to using output settings is causing trouble with the mapping of input files for certain bases. I'm not going to bother reporting it though if we're going to change this anyways.

> Just to stress it again, the preference chain is only for the master, the workers always use the "latest" setting (SRM/xrootd trumps all.)

Why have this limitation? The only

**Comment by annawoodard** [comment]
There is a limitation of course that we can't use WQ as a 'fallback' method: either you told WQ to ship the file or not. So that will have to be handled differently...

**Comment by matz-e** [comment]
For that last comment, @annawoodard: currently, there are no fallback methods (unless you changed that again), other than on the master.   As soon as chirp/srm/xrootd are specified, that's what the tasks will use.

**Comment by matz-e** [comment]
Oh and yes, I guess we should move this to after the meeting?

**Comment by annawoodard** [comment]
> For that last comment, @annawoodard: currently, there are no fallback methods (unless you changed that again), other than on the master. As soon as chirp/srm/xrootd are specified, that's what the tasks will use.

I know that's how it is, I'm voting on my opinion about how it should be :). After the meeting sounds good to me.

**Comment by klannon** [comment]
I had a revelation during my (short) commute home.  (Maybe I need a longer commute.)  We shouldn't look at transfer via WQ as another protocol, in the same way that "root:", "srm:", etc.  Transfer via WQ should be a boolean option that modifies how lobster treats the inputs.  Say we write an input specification as follows:  (excuse me if I flub the YAML):
```yaml
input:
   - file:/hadoop/store/user/lannon/ttH_discovery
   - hdfs:/store/user/lannon/ttH_discovery
   - root:T3_US_NotreDame/store/user/lannon/ttH_discovery
   - srm:T3_US_NotreDame/store/user/lannon/ttH_discovery
   - chirp://earth.crc.nd.edu:7322/ttH_discovery
want_wq_input_transfer: false
``` 
So, with `want_wq_input_transfer: false` this means that the worker should try to transfer (or copy for local files) the input to the worker.  With `want_wq_input_transfer: true` the Lobster master would transfer the input files.  Implementing the logic this way means I can take write configs that will work regardless if I run the master on the local system or remotely (e.g. at CERN for inputs located at ND).

OK, so this idea sounds so good to me, that there must be several significant flaws.  I'm sure you'll point them out tomorrow.

**Comment by annawoodard** [comment]
Here are some links from the discussion today:
* [XrootD python api](http://xrootd.org/doc/python/xrootd-python-0.1.0/examples/file.html)
* One possibility for checking XrootD (from [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookXrootdService)):

    `xrdfs cms-xrd-global.cern.ch locate /store/path/to/file`


Also, here's an idea from @klannon that came up afterwards: we could have a 'shuffle' option that would shuffle the access methods in order to achieve cheap 'load balancing' (for example, you could have multiple chirp servers.)

**Comment by matz-e** [comment]
So, to summarize:  We have a (optional) input location, an output location, and ways to access them.  Example configuration:

```yaml
storage:
  input:
    - file:///hadoop/store/user/lannon/ttH_discovery
    - hdfs:///store/user/lannon/ttH_discovery
    - root://T3_US_NotreDame/store/user/lannon/ttH_discovery
    - srm://T3_US_NotreDame/store/user/lannon/ttH_discovery
    - chirp://earth.crc.nd.edu:7322/ttH_discovery
  output:
    - hdfs:///store/user/lannon/ttH_discovery # will be used by master, ignored by worker
    - srm://T3_US_NotreDame/store/user/lannon/ttH_discovery
  use work queue for inputs: false
  use work queue for outputs: false
  disable input streaming: false
  shuffle inputs: false
  shuffle outputs: false
```

When dealing with files, these lists will be transversed in the order specified, unless shuffled.  The master will use the output access list to deal with the output SE.  The worker will probe for availability and then use the first method which grants file access.

Work plan:
 
* Rewrite the `se` module to actually try different access methods (currently is path based)
* Rewrite the `job` module to send along the input/output lists (merge jobs: these should both be identical to the output list) execution code to take
* Adjust the task wrapper to deal with the methods and fill in blanks:
  * xrootd stage-out
  * srm stage-in
  * …
* Document everything

I could take over some, or even all of the work, if needed.  It's not going to happen tonight, still, though.  But I also made the last implementation very complicated…

**Comment by annawoodard** [comment]
Sounds good! I only have one point of clarification:

>     - hdfs:///store/user/lannon/ttH_discovery # will be used by master, ignored by worker
> ...
> The master will use the output access list to deal with the output SE. The worker will probe for availability and then use the first method which grants file access.

Are you saying that worker and master behavior will be the same? If not, why? In other words, the worker doesn't have to explicitly be told to ignore `hdfs://`, it would just try it and move on to the next thing when it didn't work, the same as any other access method. Of course, if we always know something will fail, maybe it makes sense to put special handling in; on the other hand, I think we should avoid that when possible because it 1) adds complication and 2) reduces generality (for example, workers running on the T3 could use hdfs for access.)

**Comment by matz-e** [comment]
Sorry, that was misleading.  And your argument holds, behavior will be the same.  Basically, we currently don't have a way to copy in and out of hdfs except for FUSE, right?  So I would currently just skip it on the worker, until someone comes along and hacks together an implementation.  No special treatment whatsoever…

**Comment by annawoodard** [comment]
We can use hadoopy to copy in and out of hdfs, as we use it for statting files (see:
http://www.hadoopy.com/en/latest/api.html#hadoopy.put).

**Comment by klannon** [comment]
OK, this all sounds good.  (By the way, on our T3, CMSSW uses the try something even if you know it will fail thing because when you pass a LFN into cmsRun, it **always** tries to open the file locally, even though we 99% of the time want it to open it via xrootd.  As long as the test is quick, you barely notice the performance hit.)

Regardless, I think we should divide this work up in a sensible way.  @matz-e, I think you should just divide things up in whatever way minimizes global effort.  (That includes the effort involved in transferring knowledge from you to others so that they can do the work.)  To help you calibrate my usefulness: I have looked at the code for se.py (and fs.py) and I can honestly say it would take me several days of serious effort before I could understand this well enough to start changing it.  I don't think I have that much time.  However, if you could break it down a little bit, I could probably do something with guidance.  Alternatively, I'm accumulating several different odd workflows that I would be happy to use to test things, and I can easily help with documentation.

**Comment by matz-e** [comment]
@annawoodard hadoopy requires compilation, and that is something I would not want to attempt on the worker anymore.  Maybe this functionality is possible with the hadoop command line tools?

@klannon, @annawoodard I'll start on changing the master side of things.  I'll push a branch and then one of you could deal with things on the wrapper side?

**Comment by matz-e** [comment]
@klannon testing (and workflows) would be a real help.  Most of the time, I just run a small fake analysis-like workflow to test things.  I should really run the mc-production we have, too.  Anything you have that would be of general interest for testing (i.e. easy to set-up), you could put into the [examples](https://github.com/matz-e/lobster/tree/master/examples) directory.

**Comment by annawoodard** [comment]
> @annawoodard hadoopy requires compilation, and that is something I would not want to attempt on the worker anymore. Maybe this functionality is possible with the hadoop command line tools?

Ah, right. Duh. Sorry I didn't think that through. Nevermind, I think we should leave it out for now and put it in if we are seeing problems again.

> I'll push a branch and then one of you could deal with things on the wrapper side?

I'm happy to do the wrapper updates, unless @klannon has a burning desire to do it. I don't actually think we'll need too many changes (we're already handling most cases; we'll just need to change the path mapping, and switch to handling cases in succession.) Famous last words, I'm sure.

**Comment by klannon** [comment]
@matz-e: Everything I have is really derived from the examples.  The main difference is that they are based on actual analysis needs as opposed to.  Just to list them, here's what I'm picturing:
* SLHC track trigger simulation: no inputs, EDM outputs, merging required.  [Currently tested and ready to go.]
* 13 TeV multilepton tree production.  I'd like to be able to spin through the full data sets in a reasonably short time frame.  Currently this requires running on miniAOD via xrootd and putting ROOT tree outputs (from TFileService) into /hadoop/store.  Will eventually be a big list of datasets, like BEAN production.  Might eventually morph into two stages (production of skimmed/enhanced miniAOD stored locally and running over those locally to make tree output).  [Status: Testing with one dataset.  Working out issues that might be coming from new worker accounting in WQ...]
* Analysis plotting code:  For both 8 TeV and 13 TeV, want to be able to run through the DrawPlots chain using opportunistic resources so no one ever has to wait in line to make plots.  Inputs read from our local storage (or possibly copied for performance?), very small outputs (histograms) returned not to Hadoop? [Not started...]
* Running Andrew's analysis code start from BEANs through trees.  [Not started...]
* Making giant piles of 13 TeV MC.  [Not started...]

I'm happy to put any established configs into Examples with the understanding that I'm just going to put them in "as is" maybe with some comments or a README to explain how they were used.  I don't think it makes sense to try to turn them into generalized example configs though.  Does that make sense?

@annawoodard: I'm happy for you to make the wrapper-side modifications, but if you find that these aren't so quick for you and want to divide up the effort, just let me know.

**Comment by matz-e** [comment]
@klannon, some of these match what we have in `examples` already.  I would think we'd only want to ship light workflows for quick verification, and do more complex ones by experienced users (in this case you?)

I just pushed `se-tweaks` as a branch.  It passes the list through to the workers, but needs work on the master side still to really go through access methods until succession (mostly for creating directories or deleting files.)  Currently, the parameters passed to the job look like:

```json
  "mask": {
    "files": [
      "/store/mc/Summer12_DR53X/TTH_Inclusive_M-125_8TeV_pythia6/AODSIM/PU_S10_START53_V7A-v1/00000/C8FF9017-010A-E211-9A12-0025B3E05C2C.root",␣
      "/store/mc/Summer12_DR53X/TTH_Inclusive_M-125_8TeV_pythia6/AODSIM/PU_S10_START53_V7A-v1/00000/607FDFCE-100A-E211-BE93-002481E15492.root",␣
      "/store/mc/Summer12_DR53X/TTH_Inclusive_M-125_8TeV_pythia6/AODSIM/PU_S10_START53_V7A-v1/00000/72263F5C-010A-E211-A570-002590200914.root"

…

  "output": [
    "hdfs:///store/user/matze/se_test_101/",␣
    "file:///hadoop/store/user/lannon/se_test_101/",␣
    "root://xrootd.unl.edu//store/user/lannon/se_test_101",␣
    "srm://ndcms.crc.nd.edu:8443/srm/v2/server?SFN=/hadoop/store/user/lannon/se_test_101",␣
    "chirp://earth.crc.nd.edu:7322/se_test_101/"
  ],␣
  "input": [
    "file:///hadoop/store/user/lannon/se_test_101/",␣
    "hdfs:///store/user/lannon/se_test_101/",␣
    "root://xrootd.unl.edu//store/user/lannon/se_test_101",␣
    "srm://ndcms.crc.nd.edu:8443/srm/v2/server?SFN=/hadoop/store/user/lannon/se_test_101",␣
    "chirp://earth.crc.nd.edu:7322/se_test_101/"
  ]
```

Note that in this case (streaming from external resources), I don't think we should have to specify input access methods.  In case of a merge job, `input` will be using the same methods as `output`.

**Comment by klannon** [comment]
What is "mask" for?  We didn't discuss that.

**Comment by matz-e** [comment]
That's the input mask (an old parameter.)  We need to pass the required input files to the job somehow…

**Comment by annawoodard** [comment]
I've pushed what I have so far on the worker modifications, I'll finish up shortly. Here are the most important points I think that are still missing:

Worker-side:
  * Finish modifications for check_outputs
  * Handling of `hdfs://` is missing (I recommend we leave it unhandled until we see that the missing handling is causing a problem, since simpler code is easier to maintain.)
  * Testing (+ may be missing some error handling)

Master-side:
  * Handling local datasets. (To stay true to the fallback philosophy we're implementing, it seems like these should be defined relative to whatever directory is specified in config['input']. @matze, can you add to the example [here](https://github.com/matz-e/lobster/issues/143#issuecomment-124153987) to clarify what the user would put in each task's `files` specification in case he or she wanted to run on unpublished files?)

Note that for now I'm using the simplification that if we can access the file via `xrdfs` we can access it again via `cmsRun` (in other words, I'm not falling back again in case there's an error the second time). I think we should try it out this simpler way, and add something more sophisticated later if that turns out to be a problem.

**Comment by annawoodard** [comment]
Update for @matz-e: check_output modifications are in. I will keep testing, but I think most of the needed worker-side changes are there (or at least, ready for review).

**Comment by matz-e** [comment]
Seems ready from my side.  I could run this in a little while with my own workflows…

**Comment by annawoodard** [comment]
This breaks my main use mode, running over local inputs. Are you suggesting we open that in a separate issue?

**Comment by matz-e** [comment]
Ah, sorry, I was missing that one bit.  There's a transformation parameter I added a whole while ago, but that is **not** elegant at all.  I'll dig it out.

**Comment by annawoodard** [comment]
Why do we need another parameter? I thought that was the point of defining the `inputs` parameter.

**Comment by matz-e** [comment]
It's historic, and I needed it to run over the beans by prepending `/hadoop` to the information in DBS.  Currently, from what you are saying, it looks like we need match up the file information in the DB for local files to the path given in the inputs parameter?

**Comment by matz-e** [comment]
Given this:
```
storage:
  input:
    - file:///hadoop/store/user/lannon/ttH_discovery
    - hdfs:///store/user/lannon/ttH_discovery
    - root://T3_US_NotreDame/store/user/lannon/ttH_discovery
    - srm://T3_US_NotreDame/store/user/lannon/ttH_discovery
    - chirp://earth.crc.nd.edu:7322/ttH_discovery
```
and a local dataset from `file:///hadoop/store/user/lannon/ttH_discovery/eraABCDE`, it would probably be best to require to have a URL for data discovery **and** the same URL/protocol basepath in `inputs` to get files like `eraABCDE/1.root`?

**Comment by annawoodard** [comment]
>  it would probably be best to require to have a URL for data discovery and the same URL/protocol basepath in inputs to get files like eraABCDE/1.root?

I guess if we can't get around it, but it seems like it would be nice if we could use the same methods for data discovery (file, hdfs, root, chirp all support `stat` and `ls`.)

**Comment by matz-e** [comment]
I had something like that in already, just need to re-check the implementation.

**Comment by annawoodard** [comment]
The data discovery part wasn't working when I tested it, because `config['tasks']['files']` need to be defined relative to `inputs`, but in dataset.py we just try to `ls` without prepending anything.

**Comment by matz-e** [comment]
This is with the current master, which is nearly there:
```python
>>> from lobster.cmssw import dataset
>>> i = dataset.MetaInterface()
>>> i.get_info({'label': 'foo', 'files': 'srm://srm-eoscms.cern.ch:8443/srm/v2/server?SFN=/eos/cms/store/user/matze/test_se/ttH_125'}, '')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "lobster/cmssw/dataset.py", line 35, in get_info
    info = self.__file_interface.get_info(cfg)
  File "lobster/cmssw/dataset.py", line 131, in get_info
    dset.files += [f for f in fs.ls(os.path.join(entry, '*'))]
  File "lobster/se.py", line 51, in switch
    return imp.fixresult(getattr(imp, attr)(imp.lfn2pfn(path)))
  File "lobster/se.py", line 69, in fixresult
    return map(pfn2lfn, res)
  File "lobster/se.py", line 227, in ls
    for p in self.execute('ls', path).splitlines():
  File "lobster/se.py", line 201, in execute
    raise IOError(msg)
IOError: Failed to execute 'lcg-ls -b -D srmv2 srm://srm-eoscms.cern.ch:8443/srm/v2/server?SFN=/eos/cms/store/user/matze/test_se/ttH_125/*':
/eos/cms/store/user/matze/test_se/ttH_125/*: [SE][Ls][SRM_INVALID_PATH] No such file or directory.
```

**Comment by matz-e** [comment]
I guess `getsize` is also not yet implemented everywhere…

**Comment by annawoodard** [comment]
But by defining `files` hardcoded to one access method, whatever we get back (and put in the database) will always be relative to that access method, so fallbacks will fail? I was assuming we would define `files` relative to the directory pointed to by `input`.

**Comment by matz-e** [comment]
My idea behind it is that we chop of the common path by the same access methods, and then are left with the relative path as used in `inputs`.  But then again, I guess the specification for the dataset could be relative already?

**Comment by annawoodard** [comment]
Sorry, I don't really follow. Can you give an example lobster `files` and `inputs` configuration to clarify?

**Comment by annawoodard** [comment]
> I guess getsize is also not yet implemented everywhere…

I honestly think that having this information is probably not worth the extra complication. I often disable it in my local version in order to save time. We could just only collect it for DBS inputs.

**Comment by matz-e** [comment]
```yaml
storage:
  input:
    - file:///hadoop/store/user/lannon/ttH_discovery
    - hdfs:///store/user/lannon/ttH_discovery
    - root://T3_US_NotreDame/store/user/lannon/ttH_discovery
    - srm://T3_US_NotreDame/store/user/lannon/ttH_discovery
    - chirp://earth.crc.nd.edu:7322/ttH_discovery
```
and
```yaml
tasks:
  - label: foo
    files: eraABCDEFG
```
would look in `file:///hadoop/store/user/lannon/ttH_discovery/eraABCDEFG`.

Sounds good?

**Comment by annawoodard** [comment]
Sounds good!
